### PR TITLE
Allow users to configure either unstructured or JSON logging

### PR DESCRIPTION
**Pull request by crhino**
## What Does This PR Do?

This PR provides the necessary functionality for configuring all logging as either unstructured logs OR structured JSON logs.

We implement a number of different places where JSON logging can be chosen:
- `log_json: (true|false)` in the agent configuration file
- `-log-json` CLI flag on the `consul agent` command
- `-log-json` CLI flag on the `consul connect proxy` command
- `-log-json` CLI flag on the `consul monitor` command
- `logjson` query parameter on the `/v1/agent/monitor` HTTP API endpoint

```
$ consul monitor -log-level=debug -log-json
{"@level":"debug","@message":"OutgoingRPCWrapper","@module":"agent.tlsutil","@timestamp":"2020-01-23T19:26:29.295040Z","version":1}
{"@level":"debug","@message":"Rebalanced servers, new active server","@module":"agent.router.manager","@timestamp":"2020-01-23T19:26:29.297191Z","active_server":"consul-srv2.chris1 (Addr: tcp/172.20.0.2:8300) (DC: chris1)","number_of_servers":3}
{"@level":"debug","@message":"memberlist: Initiating push/pull sync with: 172.20.0.5:8301","@module":"agent.server.memberlist","@timestamp":"2020-01-23T19:26:32.362789Z"}
{"@level":"debug","@message":"memberlist: Initiating push/pull sync with: 172.20.0.2:8302","@module":"agent.server.memberlist","@timestamp":"2020-01-23T19:26:48.758943Z"}
```

In addition to these user-facing updates, we internally converted all of consul to use the `hclog` [library](https://github.com/hashicorp/go-hclog) to achieve this, converging on the same logger library as Nomad and others. Part of the internal refactoring also meant better support for key/value pairs in log output. The philosophy taken when converting log lines was remove any part of the log message that was dynamic in nature, ie. a service name, and put that into a key/value pair. This way, the number of unique messages is minimized and someone could search for specific key/value pairs.

### Miscellaneous Changes
- Rename `logger` package to `logging` so that we do not conflict with variables named `logger`. This is mostly because we now define logging name constants in `logging/names.go`.
- Added `logger/monitor` package inspired by Nomad's `monitor` [package](https://github.com/hashicorp/nomad/tree/fea44b0d8e20279a1c8fe60d4991451d34871d83/command/agent/monitor). This was to support the `consul monitor` command changes.
- Remove all references to `LevelFilter` and `LogWriter` and delete those structs. They are no longer necessary as `hclog.Logger` replaces `LevelFilter` and `monitor.Monitor` replaces `LogWriter` usage.
- Add a `loggerStore` struct in the `agent/consul` package so that the `Server` struct is able to cache already created subsystem loggers. This was useful because there are MANY different subsystems interacting in the `agent/consul` package, and the various places where the same logger wanted to be used made it fairly unwieldy to pass through the correct logger everywhere. Instead, we opted to cache loggers and initialize the logger variables in various functions.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -2385,7 +2434,7 @@ func (a *Agent) addServiceInternal(req *addServiceRequest) error {
 	if service.Proxy.Expose.Checks {
 		err := a.rerouteExposedChecks(psid, service.Proxy.LocalServiceAddress)
 		if err != nil {
-			a.logger.Println("[WARN] failed to reroute L7 checks to exposed proxy listener")
+			a.logger.Warn("to reroute L7 checks to exposed proxy listener")
```
Comment (rboyer) [inline agent/agent.go]:
```suggestion
			a.logger.Warn("failed to reroute L7 checks to exposed proxy listener")
```

File: agent/agent.go [hunk 2]
```diff
@@ -3623,14 +3740,14 @@ func (a *Agent) loadTokens(conf *config.RuntimeConfig) error {
...
-		a.logger.Printf("[WARN] unable to load persisted tokens: %v", persistenceErr)
+		a.logger.Warn("unable to load persisted tokens", "error", persistenceErr)
 	}
 
 	if persistedTokens.Default != "" {
 		a.tokens.UpdateUserToken(persistedTokens.Default, token.TokenSourceAPI)
 
 		if conf.ACLToken != "" {
-			a.logger.Printf("[WARN] \"default\" token present in both the configuration and persisted token store, using the persisted token")
+			a.logger.Warn("\"default\"token present in both the configuration and persisted token store, using the persisted token")
```
Comment (rboyer) [inline agent/agent.go]:
```suggestion
			a.logger.Warn("\"default\" token present in both the configuration and persisted token store, using the persisted token")
```

File: agent/connect/ca/provider_aws.go [hunk 3]
```diff
@@ -296,7 +300,7 @@ func (a *AWSProvider) createPCA() error {
 		},
 	}
 
-	a.logger.Printf("[DEBUG] creating new PCA %s", commonName)
+	a.logger.Debug("new PCA", "common_name", commonName)
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]:
```suggestion
	a.logger.Debug("creating new PCA", "common_name", commonName)
```

File: agent/connect/ca/provider_aws.go [hunk 4]
```diff
@@ -533,7 +536,7 @@ func (a *AWSProvider) SetIntermediate(intermediatePEM string, rootPEM string) er
 		Certificate:             []byte(intermediatePEM),
 		CertificateChain:        []byte(rootPEM),
 	}
-	a.logger.Printf("[DEBUG] uploading certificate for %s", a.arn)
+	a.logger.Debug("certificate for PCA", "pca", a.arn)
```
Comment (rboyer) [inline agent/connect/ca/provider_aws.go]:
```suggestion
	a.logger.Debug("uploading certificate for PCA", "pca", a.arn)
```

File: agent/consul/acl_endpoint.go [hunk 5]
```diff
@@ -65,26 +67,29 @@ func (a *ACL) fileBootstrapResetIndex() uint64 {
 	raw, err := ioutil.ReadFile(path)
 	if err != nil {
 		if !os.IsNotExist(err) {
-			a.srv.logger.Printf("[ERR] acl.bootstrap: failed to read %q: %v", path, err)
+			a.logger.Error("failed to read path",
```
Comment (rboyer) [inline agent/consul/acl_endpoint.go]: You've lost any indication this is related to acl bootstrapping from this error message. Probably either do a sub-logger like the original or just adjust the string message a bit?

File: agent/consul/acl_replication.go [hunk 6]
```diff
@@ -289,12 +290,10 @@ func (s *Server) updateLocalACLType(ctx context.Context, tr aclTypeReplicator) (
...
-		s.logger.Printf(
-			"[DEBUG] acl: %s replication - upserted 1 batch with %d %s of size %d",
-			tr.SingularNoun(),
-			batchEnd-batchStart,
-			tr.PluralNoun(),
-			batchSize,
+		logger.Debug(
+			"acl replication - upserted batch",
+			"number_upserted", batchEnd-batchStart,
+			"batch_size", batchSize,
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
			"batch_size", batchSize,
			"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 7]
```diff
@@ -353,28 +352,28 @@ func (s *Server) fetchACLTokens(lastRemoteIndex uint64) (*structs.ACLTokenListRe
...
 
-func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, lastRemoteIndex uint64) (uint64, bool, error) {
+func (s *Server) replicateACLType(ctx context.Context, logger hclog.Logger, tr aclTypeReplicator, lastRemoteIndex uint64) (uint64, bool, error) {
 	lenRemote, remoteIndex, err := tr.FetchRemote(s, lastRemoteIndex)
 	if err != nil {
 		return 0, false, fmt.Errorf("failed to retrieve remote ACL %s: %v", tr.PluralNoun(), err)
 	}
 
-	s.logger.Printf("[DEBUG] acl: finished fetching %s: %d", tr.PluralNoun(), lenRemote)
+	logger.Debug("finished fetching acls", "amount", lenRemote)
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
	logger.Debug("finished fetching acls", "amount", lenRemote, "item_type", tr.SingularNoun())
```

File: agent/consul/acl_replication.go [hunk 8]
```diff
@@ -400,39 +399,34 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
-		s.logger.Printf(
-			"[WARN] consul: ACL %s replication remote index moved backwards (%d to %d), forcing a full ACL %s sync",
-			tr.SingularNoun(),
-			lastRemoteIndex,
-			remoteIndex,
-			tr.SingularNoun(),
+		logger.Warn(
+			"ACL replication remote index moved backwards, forcing a full ACL sync",
+			"from", lastRemoteIndex,
+			"to", remoteIndex,
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
			"to", remoteIndex,
			"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 9]
```diff
@@ -400,39 +399,34 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
 
-	s.logger.Printf(
-		"[DEBUG] acl: %s replication - local: %d, remote: %d",
-		tr.SingularNoun(),
-		lenLocal,
-		lenRemote,
+	logger.Debug(
+		"acl replication",
+		"local", lenLocal,
+		"remote", lenRemote,
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
		"remote", lenRemote,
		"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 10]
```diff
@@ -400,39 +399,34 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
 	if res.LocalSkipped > 0 || res.RemoteSkipped > 0 {
-		s.logger.Printf(
-			"[DEBUG] acl: %s replication - deletions: %d, updates: %d, skipped: %d, skippedRemote: %d",
-			tr.SingularNoun(),
-			len(res.LocalDeletes),
-			len(res.LocalUpserts),
-			res.LocalSkipped,
-			res.RemoteSkipped,
+		logger.Debug(
+			"acl replication",
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
			"acl replication",
			"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 11]
```diff
@@ -400,39 +399,34 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
+			"skipped_remote", res.RemoteSkipped,
 		)
 	} else {
-		s.logger.Printf(
-			"[DEBUG] acl: %s replication - deletions: %d, updates: %d",
-			tr.SingularNoun(),
-			len(res.LocalDeletes),
-			len(res.LocalUpserts),
+		logger.Debug(
+			"acl replication",
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
			"acl replication",
			"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 12]
```diff
@@ -443,18 +437,16 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
 		} else if err != nil {
 			return 0, false, fmt.Errorf("failed to retrieve ACL %s updates: %v", tr.SingularNoun(), err)
 		}
-		s.logger.Printf(
-			"[DEBUG] acl: %s replication - downloaded %d %s",
-			tr.SingularNoun(),
-			lenUpdated,
-			tr.PluralNoun(),
+		logger.Debug(
+			"acl replication - downloaded updates",
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
			"acl replication - downloaded updates",
			"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 13]
```diff
@@ -443,18 +437,16 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
+			"amount", lenUpdated,
 		)
 	}
 
 	if len(res.LocalDeletes) > 0 {
-		s.logger.Printf(
-			"[DEBUG] acl: %s replication - performing deletions",
-			tr.SingularNoun(),
+		logger.Debug(
+			"acl replication - performing deletions",
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
			"acl replication - performing deletions",
			"item_type", tr.SingularNoun(),
```

File: agent/consul/acl_replication.go [hunk 14]
```diff
@@ -464,19 +456,19 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
 		if err != nil {
 			return 0, false, fmt.Errorf("failed to delete local ACL %s: %v", tr.PluralNoun(), err)
 		}
-		s.logger.Printf("[DEBUG] acl: %s replication - finished deletions", tr.SingularNoun())
+		logger.Debug("acl replication - finished deletions")
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
		logger.Debug("acl replication - finished deletions", "item_type", tr.SingularNoun())
```

File: agent/consul/acl_replication.go [hunk 15]
```diff
@@ -464,19 +456,19 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
 	}
 
 	if len(res.LocalUpserts) > 0 {
-		s.logger.Printf("[DEBUG] acl: %s replication - performing updates", tr.SingularNoun())
-		exit, err := s.updateLocalACLType(ctx, tr)
+		logger.Debug("acl replication - performing updates")
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
		logger.Debug("acl replication - performing updates","item_type", tr.SingularNoun())
```

File: agent/consul/acl_replication.go [hunk 16]
```diff
@@ -464,19 +456,19 @@ func (s *Server) replicateACLType(ctx context.Context, tr aclTypeReplicator, las
...
+		exit, err := s.updateLocalACLType(ctx, logger, tr)
 		if exit {
 			return 0, true, nil
 		}
 		if err != nil {
 			return 0, false, fmt.Errorf("failed to update local ACL %s: %v", tr.PluralNoun(), err)
 		}
-		s.logger.Printf("[DEBUG] acl: %s replication - finished updates", tr.SingularNoun())
+		logger.Debug("acl replication - finished updates")
```
Comment (rboyer) [inline agent/consul/acl_replication.go]:
```suggestion
		logger.Debug("acl replication - finished updates", "item_type", tr.SingularNoun())
```

File: agent/consul/client.go [hunk 17]
```diff
@@ -121,9 +122,12 @@ func NewClientLogger(config *Config, logger *log.Logger, tlsConfigurator *tlsuti
 		config.LogOutput = os.Stderr
 	}
 
-	// Create a logger
+	//Create a logger
```
Comment (rboyer) [inline agent/consul/client.go]:
```suggestion
	// Create a logger
```

File: agent/consul/client_serf.go [hunk 18]
```diff
@@ -75,9 +85,10 @@ func (c *Client) lanEventHandler() {
 			case serf.EventUser:
 				c.localEvent(e.(serf.UserEvent))
 			case serf.EventMemberUpdate: // Ignore
+				c.nodeUpdate(e.(serf.MemberEvent))
```
Comment (rboyer) [inline agent/consul/client_serf.go]: This is not a logging fix. Was this intended?

File: agent/consul/client_serf.go [hunk 19]
```diff
@@ -107,14 +120,33 @@ func (c *Client) nodeJoin(me serf.MemberEvent) {
 	}
 }
 
+// nodeUpdate is used to handle update events on the serf cluster
+func (c *Client) nodeUpdate(me serf.MemberEvent) {
```
Comment (rboyer) [inline agent/consul/client_serf.go]: This is not a logging fix. Was this intended?

File: agent/consul/config_replication.go [hunk 20]
```diff
@@ -117,12 +118,14 @@ func (s *Server) fetchConfigEntries(lastRemoteIndex uint64) (*structs.IndexedGen
 }
 
 func (s *Server) replicateConfig(ctx context.Context, lastRemoteIndex uint64) (uint64, bool, error) {
+	replicationLogger := s.loggers.Named(logging.Replication).Named(logging.ConfigEntry)
```
Comment (rboyer) [inline agent/consul/config_replication.go]: Doesn't this mean that you're caching the `Replication` logger but not the derived `Replication.ConfigEntry` logger and thus recreating it every replication loop?

File: agent/consul/intention_endpoint.go [hunk 21]
```diff
@@ -87,7 +89,7 @@ func (s *Intention) prepareApplyCreate(authz acl.Authorizer, entMeta *structs.En
 // intention as well as generally validating that the request is well-formed
 func (s *Intention) prepareApplyUpdate(authz acl.Authorizer, entMeta *structs.EnterpriseMeta, args *structs.IntentionRequest) error {
 	if !args.Intention.CanWrite(authz) {
-		s.srv.logger.Printf("[WARN] consul.intention: Update operation on intention %q denied due to ACLs", args.Intention.ID)
+		s.logger.Warn("Operation on intention denied due to ACLs", "intention", args.Intention.ID)
```
Comment (rboyer) [inline agent/consul/intention_endpoint.go]:
```suggestion
		s.logger.Warn("Update operation on intention denied due to ACLs", "intention", args.Intention.ID)
```

File: agent/consul/intention_endpoint.go [hunk 22]
```diff
@@ -103,7 +105,7 @@ func (s *Intention) prepareApplyUpdate(authz acl.Authorizer, entMeta *structs.En
 	// which must be true to perform any rename. This is the only ACL enforcement
 	// done for deletions and a secondary enforcement for updates.
 	if !ixn.CanWrite(authz) {
-		s.srv.logger.Printf("[WARN] consul.intention: Update operation on intention %q denied due to ACLs", args.Intention.ID)
+		s.logger.Warn("Operation on intention denied due to ACLs", "intention", args.Intention.ID)
```
Comment (rboyer) [inline agent/consul/intention_endpoint.go]:
```suggestion
		s.logger.Warn("Update operation on intention denied due to ACLs", "intention", args.Intention.ID)
```

File: agent/consul/leader.go [hunk 23]
```diff
@@ -390,7 +402,7 @@ func (s *Server) initializeLegacyACL() error {
 		if err != nil {
 			return fmt.Errorf("failed to create anonymous token: %v", err)
 		}
-		s.logger.Printf("[INFO] acl: Created the anonymous token")
+		s.logger.Info("Created the anonymous token")
```
Comment (rboyer) [inline agent/consul/leader.go]:
```suggestion
		s.logger.Info("Created the anonymous ACL token")
```

File: agent/consul/leader.go [hunk 24]
```diff
@@ -822,32 +835,40 @@ func (s *Server) startACLReplication() {
...
-type replicateFunc func(ctx context.Context, lastRemoteIndex uint64) (uint64, bool, error)
+type replicateFunc func(ctx context.Context, logger hclog.Logger, lastRemoteIndex uint64) (uint64, bool, error)
 
 // This function is only intended to be run as a managed go routine, it will block until
 // the context passed in indicates that it should exit.
 func (s *Server) runACLPolicyReplicator(ctx context.Context) error {
-	s.logger.Printf("[INFO] acl: started ACL Policy replication")
-
-	return s.runACLReplicator(ctx, structs.ACLReplicatePolicies, s.replicateACLPolicies)
+	policyLogger := s.aclReplicationLogger(structs.ACLReplicatePolicies.SingularNoun())
```
Comment (rboyer) [inline agent/consul/leader.go]: I'm sorry I missed this when I left a bazillion comments about losing your `item_type`. Will remove.

File: agent/consul/leader.go [hunk 25]
```diff
@@ -1454,7 +1498,9 @@ func (s *Server) reapTombstones(index uint64) {
 	}
 	_, err := s.raftApply(structs.TombstoneRequestType, &req)
 	if err != nil {
-		s.logger.Printf("[ERR] consul: failed to reap tombstones up to %d: %v",
-			index, err)
+		s.logger.Error("failed to reap tombstones up to index",
+			"index", index,
+			"error", err,
+		)
 	}
```
Comment (rboyer) [inline agent/consul/leader.go]: Skimming this I realized there's an unrelated bug here. Every `raftApply()` should have two error conditionals following it:
```
resp, err := s.raftApply(....blah....)
if err != nil {
  <do with err>
}
if respErr, ok := resp.(error); ok {
  <do with respErr>
}
```
This `raftApply` only has one.

File: agent/consul/logging.go [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+type loggerStore struct {
+	root hclog.Logger
+
+	// The key/value pairs that we store are (string, hclog.Logger) values
+	cache *sync.Map
```
Comment (rboyer) [inline agent/consul/logging.go]: Is there a particular reason you went with this approach over this?
```
l     sync.RWMutex
cache map[string]hclog.Logger
```

File: agent/consul/server_serf.go [hunk 27]
```diff
@@ -232,18 +241,31 @@ func (s *Server) lanNodeJoin(me serf.MemberEvent) {
 	}
 }
 
+func (s *Server) lanNodeUpdate(me serf.MemberEvent) {
```
Comment (rboyer) [inline agent/consul/server_serf.go]: This looks like an unrelated fix. Was adding this intentional?

File: agent/proxycfg/state.go [hunk 28]
```diff
@@ -357,7 +360,8 @@ func (s *state) initWatchesMeshGateway() error {
 	}, serviceResolversWatchID, s.ch)
 
 	if err != nil {
-		s.logger.Printf("[ERR] mesh-gateway: failed to register watch for service-resolver config entries")
+		s.logger.Named(logging.MeshGateway).
```
Comment (rboyer) [inline agent/proxycfg/state.go]: Did you want to use the logger cache here?

File: agent/retry_join.go [hunk 29]
```diff
@@ -52,23 +52,32 @@ func newDiscover() (*discover.Discover, error) {
...
-			servers, err := disco.Addrs(addr, logger)
+			servers, err := disco.Addrs(addr, logger.StandardLogger(&hclog.StandardLoggerOptions{
+				InferLevels: true,
+			}))
 			if err != nil {
 				if logger != nil {
-					logger.Printf("[ERR] agent: Cannot discover %s %s: %s", cluster, addr, err)
+					logger.Error("Cannot discover address",
+						"address", addr,
+						"cluster", cluster,
```
Comment (rboyer) [inline agent/retry_join.go]: Isn't this `cluster` field already present due to the `With("cluster",...)` used when the derived logger is created?

File: agent/retry_join.go [hunk 30]
```diff
@@ -52,23 +52,32 @@ func newDiscover() (*discover.Discover, error) {
...
+						"error", err,
+					)
 				}
 			} else {
 				addrs = append(addrs, servers...)
 				if logger != nil {
-					logger.Printf("[INFO] agent: Discovered %s servers: %s", cluster, strings.Join(servers, " "))
+					logger.Info("Discovered servers",
+						"cluster", cluster,
```
Comment (rboyer) [inline agent/retry_join.go]: Isn't this `cluster` field already present due to the `With("cluster",...)` used when the derived logger is created?

File: agent/retry_join.go [hunk 31]
```diff
@@ -115,15 +124,18 @@ func (r *retryJoiner) retryJoin() error {
 		return err
 	}
 
-	r.logger.Printf("[INFO] agent: Retry join %s is supported for: %s", r.cluster, strings.Join(disco.Names(), " "))
-	r.logger.Printf("[INFO] agent: Joining %s cluster...", r.cluster)
+	r.logger.Info("Retry join is supported for the following discovery methods",
+		"cluster", r.cluster,
```
Comment (rboyer) [inline agent/retry_join.go]: Isn't this `cluster` field already present due to the `With("cluster",...)` used when the derived logger is created?

File: agent/router/manager.go [hunk 32]
```diff
@@ -217,6 +218,19 @@ func (m *Manager) FindServer() *metadata.Server {
 	return l.servers[0]
 }
 
+func (m *Manager) checkServers(fn func(srv *metadata.Server) bool) bool {
```
Comment (rboyer) [inline agent/router/manager.go]: Did you mean to include this?

File: agent/router/manager.go [hunk 33]
```diff
@@ -217,6 +218,19 @@ func (m *Manager) FindServer() *metadata.Server {
...
+	for _, srv := range m.getServerList().servers {
+		if !fn(srv) {
+			return false
+		}
+	}
+	return true
+}
+
+func (m *Manager) CheckServers(fn func(srv *metadata.Server) bool) {
```
Comment (rboyer) [inline agent/router/manager.go]: Did you mean to include this?

File: agent/router/serf_adapter.go [hunk 34]
```diff
@@ -61,11 +69,12 @@ func HandleSerfEvents(logger *log.Logger, router *Router, areaID types.AreaID, s
 
 			// All of these event types are ignored.
 			case serf.EventMemberUpdate:
+				handleMemberEvent(logger, router.AddServer, areaID, e)
```
Comment (rboyer) [inline agent/router/serf_adapter.go]: Did you mean to include this?

File: api/agent.go [hunk 35]
```diff
@@ -869,12 +869,15 @@ func (a *Agent) DisableNodeMaintenance() error {
 // Providing a non-nil stopCh can be used to close the connection and stop the
 // log stream. An empty string will be sent down the given channel when there's
 // nothing left to stream, after which the caller should close the stopCh.
-func (a *Agent) Monitor(loglevel string, stopCh <-chan struct{}, q *QueryOptions) (chan string, error) {
+func (a *Agent) Monitor(loglevel string, logJSON bool, stopCh <-chan struct{}, q *QueryOptions) (chan string, error) {
```
Comment (rboyer) [inline api/agent.go]: You can't do this, the signature of the old method `Monitor` has to remain as `Monitor(string, <-chan struct{}, *QueryOptions) (chan string, error)`. You can however make a new function like `MonitorJSON`.

File: api/watch/plan.go [hunk 36]
```diff
@@ -39,22 +46,20 @@ func (p *Plan) RunWithConfig(address string, conf *consulapi.Config) error {
...
 	return p.RunWithClientAndLogger(client, logger)
 }
 
 // RunWithClientAndLogger runs a watch plan using an external client and
 // log.Logger instance. Using this, the plan's Datacenter, Token and LogOutput
 // fields are ignored and the passed client is expected to be configured as
 // needed.
-func (p *Plan) RunWithClientAndLogger(client *consulapi.Client,
-	logger *log.Logger) error {
+func (p *Plan) RunWithClientAndLogger(client *consulapi.Client, logger hclog.Logger) error {
```
Comment (rboyer) [inline api/watch/plan.go]: You can't change the signature of this method as it would break backwards compatibility in the public `api/watch` package. You'll have to make a new method.

File: command/agent/agent.go [hunk 37]
```diff
@@ -60,9 +59,13 @@ type cmd struct {
 	versionHuman      string
 	shutdownCh        <-chan struct{}
 	flagArgs          config.Flags
-	logFilter         *logutils.LevelFilter
 	logOutput         io.Writer
-	logger            *log.Logger
+	logger            hclog.InterceptLogger
+}
+
+type GatedUi struct {
```
Comment (rboyer) [inline command/agent/agent.go]: Did you mean to include this?

File: command/agent/agent.go [hunk 38]
```diff
@@ -101,18 +104,19 @@ func (c *cmd) readConfig() *config.RuntimeConfig {
 // checkpointResults is used to handler periodic results from our update checker
 func (c *cmd) checkpointResults(results *checkpoint.CheckResponse, err error) {
 	if err != nil {
-		c.UI.Error(fmt.Sprintf("Failed to check for updates: %v", err))
+
```
Comment (rboyer) [inline command/agent/agent.go]: stray newline

File: sdk/testutil/testlog.go [hunk 39]
```diff
@@ -15,12 +17,31 @@ func init() {
 	sendTestLogsToStdout = os.Getenv("NOLOGBUFFER") == "1"
 }
 
-func TestLogger(t testing.TB) *log.Logger {
```
Comment (rboyer) [inline sdk/testutil/testlog.go]: We really shouldn't delete functions from this library as someone may be using them so the change would not be backwards compatible.

You can however leave the original function intact and indicate with a comment that it is deprecated:

```
// Deprecated: use Logger(t)
func TestLogger(t testing.TB) *log.Logger {
```
see: https://rakyll.org/deprecated/

File: sdk/testutil/testlog.go [hunk 40]
```diff
@@ -15,12 +17,31 @@ func init() {
...
+
+func LoggerWithOutput(t testing.TB, output io.Writer) hclog.InterceptLogger {
+	return hclog.NewInterceptLogger(&hclog.LoggerOptions{
+		Name:   t.Name(),
+		Level:  hclog.Debug,
+		Output: output,
+	})
 }
 
-func TestLoggerWithName(t testing.TB, name string) *log.Logger {
```
Comment (rboyer) [inline sdk/testutil/testlog.go]: Similarly for backwards compatibility we should retain the old method and just mark it deprecated:

```
// Deprecated: use LoggerWithName(t)
func TestLoggerWithName(t testing.TB, name string) *log.Logger {
```

File: agent/consul/acl_endpoint.go (hunk 5 above)
Comment (s-christoff) [reply agent/consul/acl_endpoint.go]: It does output `agent.acl` - but I can add  "bootstrap" to the front of it - is that what you're requesting?

File: agent/consul/client_serf.go (hunk 18 above)
Comment (s-christoff) [reply agent/consul/client_serf.go]: We must of pulled something we shouldn't have. :|

File: agent/consul/config_replication.go (hunk 20 above)
Comment (s-christoff) [reply agent/consul/config_replication.go]: Yes - but I'm not finding a _cleaner_ way to do this. Any suggestions?

File: agent/consul/leader.go (hunk 23 above)
Comment (s-christoff) [reply agent/consul/leader.go]: This will print out "agent.acl Info Create the anonymous ACL token", so this is already prefaced with ACL but I'm not opposed to adding this - just want to clarify

File: agent/retry_join.go (hunk 29 above)
Comment (s-christoff) [reply agent/retry_join.go]: Yeah, so this will output `agent.cluster [LOGLEVEL] cannot discovery address cluster LAN/WAN`, the way that it is outputting will allow users to be able to search for the key cluster and parse by LAN/WAN if they want

Comment (rboyer) [reply agent/retry_join.go]: I meant that you were using `With()` to already give its sublogger the `cluster` param:
```
// Creates a sublogger that will always have the given key/value pairs
With(args ...interface{}) Logger
```

File: agent/consul/leader.go (hunk 23 above)
Comment (rboyer) [reply agent/consul/leader.go]: Right. Nevermind.

File: agent/consul/config_replication.go (hunk 20 above)
Comment (rboyer) [reply agent/consul/config_replication.go]: in `agent/consul/server.go` make `logger` a derived logger:
```
configReplicatorConfig := ReplicatorConfig{
		Name:     "Config Entry",
		Name:     logging.ConfigEntry,
		Delegate: &FunctionReplicator{ReplicateFn: s.replicateConfig},
		Rate:     s.config.ConfigReplicationRate,
		Burst:    s.config.ConfigReplicationBurst,
		Logger:   logger,
	}
```

in fact for any of those other `ReplicatorConfig` structs you might want to use item-type based subloggers for all of them.

File: agent/consul/acl_endpoint.go (hunk 5 above)
Comment (s-christoff) [reply agent/consul/acl_endpoint.go]: * will re-add bootstrap

File: agent/retry_join.go (hunk 29 above)
Comment (s-christoff) [reply agent/retry_join.go]: *remove cluster k/v

File: agent/consul/leader.go (hunk 25 above)
Comment (s-christoff) [reply agent/consul/leader.go]: I think because with this we are requesting raft do something, and we don't have to do anything but hope it doesn't give us an err?

File: agent/consul/config_replication.go (hunk 20 above)
Comment (rboyer) [reply agent/consul/config_replication.go]: Did you want to use this new `logger` function parameter instead of the `replicationLogger` variable now?

File: agent/consul/logging.go [hunk 41]
```diff
@@ -0,0 +1,31 @@
...
+
+func (ls *loggerStore) Named(name string) hclog.Logger {
+	ls.l.Lock()
+	defer ls.l.Unlock()
+	l, ok := ls.cache[name]
+	if !ok {
+		l = ls.root.Named(name)
+		ls.cache[name] = l
+	}
+	return l.(hclog.Logger)
```
Comment (rboyer) [inline agent/consul/logging.go]:
```suggestion
	return l
```

File: api/agent.go [hunk 42]
```diff
@@ -875,6 +875,54 @@ func (a *Agent) Monitor(loglevel string, stopCh <-chan struct{}, q *QueryOptions
...
+				logCh <- ""
+			}
+		}
+	}()
+
+	return logCh, nil
+}
+
+//MonitorJSON sets the logJSON flag to true
+func (a *Agent) MonitorJSON(loglevel string, logJSON bool, stopCh <-chan struct{}, q *QueryOptions) (chan string, error) {
```
Comment (rboyer) [inline api/agent.go]: I think you might be better suited here by having 3 functions, rather than two of them with nearly identical code.
```
func (a *Agent) Monitor(_ string, _ bool, _ <-chan struct{}, _ *QueryOptions) (chan string, error) {
  /* just call monitor(..., false)*/
}
func (a *Agent) MonitorJSON(_ string, _ bool, _ <-chan struct{}, _ *QueryOptions) (chan string, error) {
  /* just call monitor(..., true)*/
}
func (a *Agent) monitor(_ string, bool, _ <-chan struct{}, _ *QueryOptions, logJSON bool) (chan string, error) {
  /* all the code */
}
```

Comment (rboyer) [reply api/agent.go]: Also in the version you have here now you still have the boolean param in the MonitorJSON variation which isn't necessary since the truthiness is in the function name

File: api/watch/plan.go [hunk 43]
```diff
@@ -39,22 +47,92 @@ func (p *Plan) RunWithConfig(address string, conf *consulapi.Config) error {
...
+				watchLogger.Error("Handler only supports index-based " +
+					" watches but non index-based watch run. Skipping Handler.")
+			}
+			p.Handler(uint64(idx), result)
+		}
+	}
+	return nil
+}
+
+//Deprecated: Use RunwithClientAndHclogger
```
Comment (rboyer) [inline api/watch/plan.go]: In the comment you use one name `RunwithClientAndHclogger` and the function definition is `RunWithClientAndHclog`. Which one do you want?

**Comment by mkcp** [comment]
https://www.youtube.com/watch?v=OFr74zI1LBM
